src/ipa/rpi/controller/rpi/clock_recovery.cpp

```diff
@@ -0,0 +1,121 @@
+#include "clock_recovery.h"
+
+#include <cctype>
+#include <chrono>
+#include <fcntl.h>
+#include <list>
+#include <map>
+#include <strings.h>
+#include <unistd.h>
+#include <vector>
+
+#include <libcamera/base/log.h>
+
+#include "sync_status.h"
+
+using namespace std::chrono_literals;
+using namespace RPiController;
+using namespace libcamera;
+
+LOG_DEFINE_CATEGORY(RPiClockRec)
+
+
+float ClockRecovery::slope()
+{
+	//always just add the last two ones, and remove before in adding fucntion
+	double sl = (error_values.size() * xysum - xsum * ysum) / (error_values.size() * x2sum - xsum * xsum);
+	if (sl) {
+		return sl;
+	} else {
+		return 0;
+	}
+}
+
+void ClockRecovery::updating_values(std::chrono::microseconds correction)
+{
+	std::list<int64_t>::iterator iterror = error_values.begin();
+	std::list<int64_t> update;
+
+	for (; iterror != error_values.end(); iterror++) {
+		update.push_back((*iterror) - correction.count());
+ 	}
+	ysum -= (double)correction.count() * update.size();
+	xysum -= correction.count() * xsum;
+	error_values = update;
+}
+
+std::chrono::microseconds ClockRecovery::trending_error(std::chrono::microseconds LastWallClock, std::chrono::microseconds ClientWallClock, std::chrono::microseconds lastPayloadFrameDuration, unsigned int sequence)
+{
+	//local wall clock - server wall clock, so if we hurry, we get positive, which shortens the frame
+	//std::chrono::microseconds lastPayloadFrameDuration = (NextWallClock - LastWallClock) /syncPeriod ;
+	std::chrono::microseconds delta = (ClientWallClock) - (LastWallClock);
+	unsigned int mul = (delta + lastPayloadFrameDuration / 2) / lastPayloadFrameDuration;
+	std::chrono::microseconds delta_mod = delta - mul * lastPayloadFrameDuration;
+	
+	int y = delta_mod.count();
+	int x = sequence;
+
+	if (error_values.size() == listSize) {
```

Comment:
Maybe a comment that you're using rolling sums would help other people to understand here?